python/pyarrow/dataset.py

```diff
@@ -232,104 +228,261 @@ def _ensure_format(obj):
+                file_info = None
+    elif isinstance(filesystem, str):
+        # instantiate the file system from an uri, if the uri has a path
+        # component then it will be treated as a path prefix
+        filesystem, prefix = FileSystem.from_uri(filesystem)
+        prefix = _normalize_path(filesystem, prefix)
+        if prefix:
+            filesystem = SubTreeFileSystem(prefix, filesystem)
+    elif not isinstance(filesystem, FileSystem):
         raise TypeError(
-            "Expected a path-like or DatasetFactory, got {}".format(type(src))
+            '`filesystem` argument must be a FileSystem instance or a valid '
+            'file system URI'
         )
 
+    # retrieve the file descriptor if it is available already
+    if file_info is None:
+        file_info = filesystem.get_file_info([path])[0]
+
+    # depending on the path type either return with a recursive
+    # directory selector or as a list containing a single file
+    if file_info.type == FileType.Directory:
+        paths_or_selector = FileSelector(path, recursive=True)
+    elif file_info.type == FileType.File:
+        paths_or_selector = [path]
+    else:
+        raise FileNotFoundError(path)
+
+    return filesystem, paths_or_selector
 
-def dataset(paths_or_factories, filesystem=None, partitioning=None,
-            format=None, schema=None):
+
+def _filesystem_dataset(source, schema=None, filesystem=None,
+                        partitioning=None, format=None,
+                        partition_base_dir=None, exclude_invalid_files=None,
+                        ignore_prefixes=None):
+    """
+    Create a FileSystemDataset which can be used to build a Dataset.
+
+    Parameters are documented in the dataset function.
+
+    Returns
+    -------
+    FileSystemDataset
+    """
+    format = _ensure_format(format or 'parquet')
+    partitioning = _ensure_partitioning(partitioning)
+
+    if isinstance(source, (list, tuple)):
+        fs, paths_or_selector = _ensure_multiple_sources(source, filesystem)
+    else:
+        fs, paths_or_selector = _ensure_single_source(source, filesystem)
+
+    options = FileSystemFactoryOptions(
+        partitioning=partitioning,
+        partition_base_dir=partition_base_dir,
+        exclude_invalid_files=exclude_invalid_files,
+        ignore_prefixes=ignore_prefixes
+    )
+    factory = FileSystemDatasetFactory(fs, paths_or_selector, format, options)
+
+    return factory.finish(schema)
+
+
+def _union_dataset(source, schema=None, **kwargs):
+    if any(v is not None for v in kwargs.values()):
+        raise ValueError(
+            "When passing a list of Datasets, you cannot pass any additional "
+            "arguments"
+        )
+
+    if schema is None:
+        # unify the children datasets' schemas
+        schema = pa.unify_schemas([ds.schema for ds in source])
+
+    # create datasets with the requested schema
+    children = [ds.replace_schema(schema) for ds in source]
```

Comment:
This logic is already applied in C++ if no schema is passed to CUnionDataset.Make

Reply:
Great!

Reply:
Adding tests and updating.

Reply:
@bkietz without that the tests are failing.